File: backend/src/main/java/com/plugga/backend/controller/UserDeckController.java
package com.plugga.backend.controller;

import com.plugga.backend.entity.DeckCard;
import com.plugga.backend.entity.UserDeck;
import com.plugga.backend.entity.UserDeck;
import com.plugga.backend.service.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/user_decks")
public class UserDeckController {
    private UserDeckService userDeckService;
    private DeckService deckService;
    private UserService userService;

    @Autowired
    public UserDeckController(UserDeckService userDeckService, DeckService deckService, UserService userService){
        this.userDeckService = userDeckService;
        this.deckService = deckService;
        this.userService = userService;
    }

    @GetMapping("/")
    public List<UserDeck> findAll(){
        return userDeckService.findAll();
    }

    @GetMapping("/{userDeckId}")
    public UserDeck getUserDeck(@PathVariable int userDeckId){
        UserDeck userDeck = userDeckService.findById(userDeckId);
        if (userDeck == null){
            throw new RuntimeException("Could not find userDeck using id: " + userDeckId);
        }
        return userDeck;
    }

    @GetMapping(value = "", params = "deckId")
    public List<UserDeck> getByDeckId(@RequestParam("deckId") int deckId){
        List<UserDeck> userDecks = userDeckService.findByDeckId(deckId);
        if (userDecks == null){
            throw new RuntimeException("Could not find userDecks using deckId: " + deckId);
        }
        return userDecks;
    }

    @GetMapping(value = "", params = "userId")
    public List<UserDeck> getByUserId(@RequestParam("userId") int userId){
        List<UserDeck> userDecks = userDeckService.findByUserId(userId);
        if (userDecks == null){
            throw new RuntimeException("Could not find userDecks using cardId: " + userId);
        }
        return userDecks;
    }

    @PostMapping("/")
    public UserDeck addUserDeck(@RequestBody UserDeck userDeck){
        userDeckService.save(userDeck);
        userDeck.setDeck(deckService.findById(userDeck.getDeck().getId()));
        userDeck.setUser(userService.findById(userDeck.getUser().getId()));
        return userDeck;
    }

//    @PutMapping("/")
//    public DeckCard updateDeckCard(@RequestBody DeckCard deckCard){
//        deckCardService.save(deckCard);
//        deckCard.setDeck(deckService.findById(deckCard.getDeck().getId()));
//        deckCard.setCard(cardService.findById(deckCard.getCard().getId()));
//        deckCard.setPile(pileService.findById(deckCard.getPile().getId()));
//        return deckCard;
//    }
//
//    @DeleteMapping("/{deckCardId}")
//    public String deleteDeckCard(@PathVariable int deckCardId){
//        DeckCard tempDeckCard = deckCardService.findById(deckCardId);
//        if (tempDeckCard == null){
//            throw new RuntimeException("Could not find deckCard using id: " + deckCardId);
//        }
//        deckCardService.deleteById(deckCardId);
//        return "Deleted deckCard with id: " + deckCardId;
//    }
}

File: backend/src/main/kotlin/com/plugga/backend/controller/UserDeckController.kt
package com.plugga.backend.controller

import com.plugga.backend.entity.UserDeck
import com.plugga.backend.service.DeckService
import com.plugga.backend.service.UserDeckService
import com.plugga.backend.service.UserService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Pageable
import org.springframework.data.web.PagedResourcesAssembler
import org.springframework.hateoas.EntityModel
import org.springframework.hateoas.PagedModel
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/user_decks")
class UserDeckController @Autowired
constructor(private val userDeckService: UserDeckService, private val deckService: DeckService, private val userService: UserService) {

    @GetMapping("")
    fun findAll(pageable: Pageable, pagedResourcesAssembler: PagedResourcesAssembler<UserDeck>): PagedModel<EntityModel<UserDeck>> {
        return pagedResourcesAssembler.toModel(userDeckService.findAll(pageable))
    }

    @GetMapping("/{userDeckId}")
    fun getUserDeck(@PathVariable userDeckId: Int): UserDeck {
        return userDeckService.findById(userDeckId)
            ?: throw RuntimeException("Could not find userDeck using id: $userDeckId")
    }

    @GetMapping(value = [""], params = ["deckId"])
    fun getByDeckId(pageable: Pageable, pagedResourcesAssembler: PagedResourcesAssembler<UserDeck>, @RequestParam("deckId") deckId: Int): PagedModel<EntityModel<UserDeck>> {
        return pagedResourcesAssembler.toModel(userDeckService.findByDeckId(pageable, deckId))
    }

    @GetMapping(value = [""], params = ["userId"])
    fun getByUserId(pageable: Pageable, pagedResourcesAssembler: PagedResourcesAssembler<UserDeck>, @RequestParam("userId") userId: Int): PagedModel<EntityModel<UserDeck>> {
        return pagedResourcesAssembler.toModel(userDeckService.findByUserId(pageable, userId))
    }

    @PostMapping("")
    fun addUserDeck(@RequestBody userDeck: UserDeck): UserDeck {
        userDeck.id = 0
        userDeckService.save(userDeck)
        userDeck.deck = deckService.findById(userDeck.deck!!.id)
        userDeck.user = userService.findById(userDeck.user!!.id)
        return userDeck
    }

    @PutMapping("/")
    fun updateUserDeck(@RequestBody userDeck: UserDeck): UserDeck {
        userDeckService.save(userDeck)
        userDeck.deck = deckService.findById(userDeck.deck!!.id)
        userDeck.user = userService.findById(userDeck.user!!.id)
        return userDeck
    }

    @DeleteMapping("/{userDeckId}")
    fun deleteUserDeck(@PathVariable userDeckId: Int): String {
        userDeckService.findById(userDeckId) ?: throw RuntimeException("Could not find userDeck using id: $userDeckId")
        userDeckService.deleteById(userDeckId)
        return "Deleted userDeck with id: $userDeckId"
    }
}
